src/engine/bufferscalers/rubberbandwrapper.cpp

```diff
@@ -0,0 +1,168 @@
+#include "engine/bufferscalers/rubberbandwrapper.h"
+
+#include "engine/bufferscalers/rubberbandworkerpool.h"
+#include "engine/engine.h"
+#include "util/assert.h"
+
+using RubberBand::RubberBandStretcher;
+
+#define RUBBERBANDV3 (RUBBERBAND_API_MAJOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
+
+namespace {
+
+mixxx::audio::ChannelCount getChannelPerWorker() {
+    RubberBandWorkerPool* pPool = RubberBandWorkerPool::instance();
+    return pPool ? pPool->channelPerWorker() : mixxx::kEngineChannelCount;
+}
+} // namespace
+
+int RubberBandWrapper::getEngineVersion() const {
+    VERIFY_OR_DEBUG_ASSERT(isValid()) {
+        return -1;
+    }
+#if RUBBERBANDV3
+    return m_pInstances[0]->getEngineVersion();
+#else
+    return 2;
+#endif
+}
+void RubberBandWrapper::setTimeRatio(double ratio) {
+    for (auto& stretcher : m_pInstances) {
+        stretcher->setTimeRatio(ratio);
+    }
+}
+size_t RubberBandWrapper::getSamplesRequired() const {
+    size_t require = 0;
+    for (auto& stretcher : m_pInstances) {
+        require = qMax(require, stretcher->getSamplesRequired());
+    }
+    return require;
+}
+int RubberBandWrapper::available() const {
+    int available = std::numeric_limits<int>::max();
+    for (auto& stretcher : m_pInstances) {
+        available = qMin(available, stretcher->available());
+    }
+    return available == std::numeric_limits<int>::max() ? 0 : available;
+}
+size_t RubberBandWrapper::retrieve(float* const* output, size_t samples) const {
+    if (m_pInstances.size() == 1) {
+        return m_pInstances[0]->retrieve(output, samples);
+    } else {
+        size_t ret = 0;
+        for (const auto& stretcher : m_pInstances) {
+            size_t thisRet = stretcher->retrieve(output, samples);
+            DEBUG_ASSERT(!ret || thisRet == ret);
+            ret = qMax(thisRet, ret);
+            output += getChannelPerWorker();
+        }
+        return ret;
```

Comment:
This will probably cause audible artifacts when one stretcher returns less data than the other ones. This is usually not a problem with rubberband because we know how much data has actually been returned, but in this case we take the max of all stretchers which will cause mixxx to read the unwritten part of the buffer. 
```
                    v max of all retrieved data
C1|#################|
C2|#########--------| <-- less data returned
            ^^^^^^^^ unwritten data (zero or leftover from previous call?)
```

Simply truncating is not an option either as the remaining data from the stretcher will then be lost (unless we manually buffer it (complicated to get right)).
```
            vvvvvvvv all the data that has been lost because its out of the stretcher, but not consumed by the engine.
C1|#################|
C2|#########--------| <-- less data returned
           ^ min of all retrieved data
```

So the only remaining solution is to query the minimum beforehand, and then only retrieve that minimum available. 

```suggestion
        // ensure we don't fetch more samples than we really have available.
        samples = std::min{available(), samples};
        for (const auto& stretcher : m_pInstances) {
            size_t numSamplesRequested = stretcher->retrieve(output, samples);
            // there is something very wrong if we got less samples
            // than were available.
            DEBUG_ASSERT(numSamplesRequested == samples);
            output += getChannelPerWorker();
        }
        return samples;
```

Reply:
This function should always be called with the value returned by `available`, which is adjusted to the minimum available. For safety, I could call it again in this method to make sure that the user isn't using an arbitrary value, but we assume that callers know what they are doing and will use a value between 0..<ReturnedAvailable> 
I think we need to remember we are not making a public library here, but a wrapper only mean to supersede the bare RubberBand instance used in the scaler, which already doesn't something along the line of `std::min{available(), samples}`

Reply:
https://github.com/acolombier/mixxx/blob/05df96d3ca385dade90caf88187355178870ce71/src/engine/bufferscalers/enginebufferscalerubberband.cpp#L143

Reply:
> This function should always be called with the value returned by available, which is adjusted to the minimum available. For safety, I could call it again in this method to make sure that the user isn't using an arbitrary value, but we assume that callers know what they are doing and will use a value between 0..

Well this is essentially a wrapper, so adhering to the liskov substition principle, the behavior needs to match. The rubberband API doesn't assume you only call it with the `available` samples so this wrapper API that looks the exact same. In the case of rubberband, its only a bug if you set `samples` to some value thats larger than the retrieval buffer, in the current iteration it would be bug to set it to something larger than available because of the previous two issues I was describing. Lets adhere to the rubberband API and only assume that `samples` is representing the buffer size. 

> I think we need to remember we are not making a public library here, but a wrapper only mean to supersede the bare RubberBand instance used in the scaler, which already doesn't something along the line of std::min{available(), samples}

That might be true in this specific case, but in a codebase as large as mixxx, every interface is essentially a library, so in order for the consumers of that interface to stay sane, it still needs to be reasonably well designed.

Reply:
I'll add a `DEBUG_ASSERT`